### ISSUE TITLE
set up github actions for building/testing/containers

### ISSUE THREAD

**Issue by joernu76** [Open-MSS/MSS#619]
I think as we have the means, the next, most important step would be to set up the github actions to automatically execute and check the syntax and test cases. Second would be to visualize coverage using the Pages feature. 
It is very wasteful that we have not automated that.

I propose this course of action in this order

1) set up an "action" to install a conda environment and execute tests in a debian (or similar) based docker container. (I do not use the conda build mechanism, normally; typically installing the proper pacakges and executing pytest should work).

2) create a container with everything neede preinstalled, and use this one instead of creating a new environment everytime

3) Replicate the conda build process, install the package and test this one.

4)  build a container based on 2) including mss

2/4 would be new projects within the Open-MSS group.

What do you think? @ReimarBauer @Marilyth

**Comment by joernu76** [comment]
Ancillarily, 2) allows us to run tests against a known-good environment as well as running (e.g. weekly) against a container employing the most recent packages, which will allow us to determine quickly if new packages break MSS.
I have a similar setup for juwave, which could serve as a pattern.

**Comment by Marilyth** [comment]
I didn't yet get to learn much about GitHub actions and how it works with containers.
I will play around with it during the weekend and see if I can make an informed opinion afterwards.

**Comment by joernu76** [comment]
Perfect. Maybe Reimar can chime in with some ideas of his. But this should be prioritized IMO, because having that set up properly will save us quite a lot of work. I have become so used to that from my other projects that checking manually MSS tests fells like walking through a quagmire.

**Comment by ReimarBauer** [comment]
I think we should manage the installation for the tests by a 
`mamba install --file requiremtents.d/base.txt` and an additional `mamba install --file requiremtents.d/development.txt` here.
This is related to https://github.com/Open-MSS/MSS/issues/612 
On future gsoc projects I do prefer in this namespace an additional requirements file, which we can remove after merge. 

I have a container on the dockerhub, but I would prefer at least a weekly rebuild of the one used for tests. Because of the good idea not to fixate to much packages, we should quite fast see that we can't use "the newest" package of something. 

@plant99 wanted to look at a containerd definition https://github.com/Open-MSS/MSS/issues/600 

I'm not sure about the effort to create the package same way as on conda-forge, I think that this is too much.

**Comment by Marilyth** [comment]
GitHub runners, as far as I am aware, just run on a console environment.
How would we go about testing the UI segments of MSS? The tests containing UI cause pytest to abort on the GitHub runners.
Perhaps we need a self-hosted runner for that.

For reference: https://github.com/Marilyth/MSS/runs/1681236755?check_suite_focus=true

**Comment by joernu76** [comment]
If we have full root rights, we can install a framebuffer-based X server and the pyvirtualdisplay module to test all the UI on the console.
I.e. I do not see an issue for that if we can install both Linux packages and Python modules as we should be able to in a docker environment.
I set my laptop up for this last Friday. The (a?) necessary debian package was 'xvfb' and the conda package was 'pyvirtualdisplay'. With both thingies installed, it should work on "a console" (by doing the UI stuff in some invisible buffer in the background).

This is also a life saver for running the tests locally as the opening and closing windows make working on something in parallel impossible. I deactivate that only for debugging purposes, when I run or create a specific, single testcase.

**Comment by ReimarBauer** [comment]
from our development docs

On linux install the conda package pyvirtualdisplay and xvfb from your linux package manager. This is used to run tests on a virtual display. If you don’t want tests redirected to the xvfb display just setup an environment variable:

$ export TESTS_VISIBLE=TRUE

**Comment by ReimarBauer** [comment]
@joernu76 I'm not sure if 3) is easily doable, or how it should be done. We should not work on a doubled conda-forge just for mss.

A very simple option is to create regularly new builds or at least send build requests and use by that the conda-forge environment.
With this we have to define tests to find package conflicts too.  To find if there are new packages available, we can just build and use the conda list output.

**Comment by joernu76** [comment]
ad 3) Me neither. However doing something like `conda build whatever` to trigger a build locally should not be very hard? Feeding the result of conda-forge back into our project here seems way more difficult.

**Comment by ReimarBauer** [comment]
I got today to an other option for the dependency check for a released stable e.g. 2.0.0 by https://docs.docker.com/docker-hub/builds/automated-testing/

this sounds like ongoing builds and wait until a build fails

**Comment by ReimarBauer** [comment]
For 2)

Lets look on this as a multidimensional problem to figure if a dependency gets broken.

We need an ongoing installation of the released version and running tests. This could be solved by dockers automated-testing.
At the time this fails, we know we have to do a new build on conda-forge with a fixation of packages.
With the new build the problem of stable is solved.

After the build of stable the changed meta.yaml gets merged into develop and all problems those modules created are also gone.
Any new build of develop based on the rules of stable will only show failures for packages we don't have in stable.
We can do the same automated docker build test for develop by building the package on our own. By this we will see which new module needs a fixation.
If that automated test fails we do the fixation of the module or refactor in develop. This has than priority. After this is fixed, we check open PR if they still fail.


Now lets look on which docker image we can use for testing stable for running tests on github. We use the last released testing-stable. Because we know when it gets broken.
And for develop we use a testing-develop for running tests on github. This has to become created at the moment we change the package list. After a release this is identical with testing-stable.


Lets look on a failing test by a PR. In stable we check if dockers automated-testing has also failed. If this also fails we do a the modules fixation and release a new build. Afterwards we have a new
docker image for tests and look again on the PR.

In develop we have a failing test on a PR. We check dockers automated-testing of the build. If this didn't fail the PR is the culprit. If it fails we have to figure what changed and this gets priority.
After we solved this the PR will be rechecked.  Also we can destinguish by looking on the failing tests and the dependency change involved if this is in relation. Because develop is not released we can also decide to merge first and look afterwards what needs to be done. Merging something broken should not be the default.

**Comment by Marilyth** [comment]
This is also more or less what I envisioned.

Now that we have the Dockerfiles for everything ready (I think), we should see to their automatic creation and testing.
After everything is automated on dockerhub, this whole endeavor will be a lot easier to understand and achieve.

**Comment by Marilyth** [comment]
So, to the best of my understanding:
1. is done already with the current action
2. this image is automatically created each day, but currently still pushed to marilyth/mss-test instead of yourmss/mss-test
    once it is pushed to the correct account, this can replace 1.
3. is automatically done as well, by creating and testing a develop docker image every day
4. I am not entirely sure what this accomplishes/tests, after all the steps before. 
    We have a latest image containing mss from 3., and we have a latest image containing only the dependencies from 2.
    It seems to me that 4. is already a result of mixing 2. and 3.

Also, all the images on dockerhub are guaranteed to work, so we should never run into dependency issues here.
If new packages break it, it will show on https://github.com/Open-MSS/dockerhub/actions instead.

**Comment by joernu76** [comment]
I think you started with 4). This serves people who want to use mss instead of just testing it :-)
From what I got during the meetings you have covered these basic steps for the most part. What we could not agree on last Thursday was the following. How to:

a) Periodically identify breakage of stable MSS due to environmental (i.e. dependency) breakage
b) ensure that the tests of a merge request to develop or stable does *not* fail due to dependency problems. 

I think that a) is adressed in some fashion by dockerhub...? How to handle b) is either way the more complex task.

**Comment by Marilyth** [comment]
a) is definitely addressed by dockerhub right now. Every day stable MSS is tested, and if the tests fail it is a sign of an environmental issue (or #624).

b) unless the merge request specifically modifies the dependencies, the dependencies won't be an issue, as the images with which the tests are run are up to date and working.
If they do modify the dependencies, we just need to install the new or updated ones in the action before the tests. This however will not guarantee no dependency issues as they are new.
